### ISSUE TITLE
external-check could not access system environments

### ISSUE THREAD

**Issue by jvincze84** [haproxy/haproxy#2128]
### Detailed Description of the Problem

I would like to use HAProxy for TCP load balancing between postgres nodes.
HAProxy is running iside a docker container.
I want to user system evironment inside my external-check bash shell script.

**Relevant Configurations:**

```
global
        log stdout format iso    local0
        insecure-fork-wanted
        insecure-setuid-wanted
        external-check

frontend psqlfe
  bind *:5432
  mode tcp
  log global
  option tcplog
  default_backend pgsql


backend pgsql
        mode tcp
        external-check path "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        external-check command /var/lib/haproxy/health-check.sh
        option external-check
        server pg-0 pg-0:5432 check inter 1s
        server pg-1 pg-1:5432 check inter 1s
```

Docker Compose:
```
    environment:
      - PG_USER=postgres
      - PGPASSWORD=adminpassword
      - PG_DB=postgres
```

Check Script:
```
#!/bin/bash

echo "PG_DB  - $PG_DB"

VIRT_IP=$3
VIRT_PORT=$4
REAL_IP=$3

if [ "$4" == "" ]; then
  REAL_PORT=$VIRT_PORT
else
  REAL_PORT=$4
fi

STATUS=$(psql -qtAX -c "select pg_is_in_recovery()" -h "$VIRT_IP" -p "$REAL_PORT" --dbname="$PG_DB" --username="$PG_USER")

if [ "$STATUS" == "f" ]; then
  # We are in master mode
  exit 0
else
  exit 1
fi
```

HAProxy log:

```
postgres-cluster_ha-proxy.1.12fm6nlo8rtf@itdevart1    | PG_DB  -
postgres-cluster_ha-proxy.1.12fm6nlo8rtf@itdevart1    | PG_DB  -
````

### Expected Behavior

External health check script should access system environment.

### Steps to Reproduce the Behavior

Try to access system environemnt from external shell script.

### Do you have any idea what may have caused this?

Maybe the external health check command run in chroot. If this is the problem, is there any workaround for this?

### What is your configuration?

```haproxy
global
        log stdout format iso    local0
        insecure-fork-wanted
        insecure-setuid-wanted
        external-check

frontend kubeapi
  bind *:5432
  mode tcp
  log global
  option tcplog
  default_backend pgsql


backend pgsql
        mode tcp
        external-check path "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        external-check command /var/lib/haproxy/health-check.sh
        option external-check
        server pg-0 pg-0:5432 check inter 1s
        server pg-1 pg-1:5432 check inter 1s
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.7.6-4dadaaa 2023/03/28 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.6.html
Running on: Linux 5.10.0-20-amd64 #1 SMP Debian 5.10.158-2 (2022-12-13) x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 -BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN +SLZ -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=4).
Built with OpenSSL version : OpenSSL 3.0.8 7 Feb 2023
Running on OpenSSL version : OpenSSL 3.0.8 7 Feb 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.1 20220924

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by wtarreau** [comment]
Indeed I looked at the code and I'm seeing that a fresh new environment is created for executed commands, I didn't remember about this. I'm wondering if we could have an option to merge the new variables and the current environment before launching the command. An alternate solution could be to have a list of explicit variables to declare but that could easily be annoying.

You may want to give this a try. Warning: I have not tested it at all, and it could even fail to build. It tries to scan all the check variables and append them to the existing environment. It's not configurable bit it could be a good start.

[env.diff.txt](https://github.com/haproxy/haproxy/files/11290795/env.diff.txt)

**Comment by git001** [comment]
I would vote for a optin like this.

```
external-check import-env ...
```

and there must be at least one argument.
In the case here would the line looks like this.

```
external-check import-env PG_DB PG_USER
```

**Comment by wtarreau** [comment]
I initially thought about something like this but figured that when users really need environment, it means they might be dealing with complicated tools or advanced scripts that rely on many variables. When this happens, it becomes a pain to enumerate all variables, so I think that we should either import all the environment or none of it. In this case we could use "external-check import-env" to mention that it imports all the environment, and be done with it.

**Comment by nyxi** [comment]
Support for this would really make external-checks much more powerful.

It is not just the system environment variables that are ignored, but also anything you set directly in the haproxy config with `setenv`/`presetenv` which I would expect to be passed on to any subprocess.

**Comment by wtarreau** [comment]
Then could you please try the proposed patch above ? I proposed it so as to get some feedback, but I would like it to be tested before deciding to merge it. Thanks!

**Comment by nyxi** [comment]
> Then could you please try the proposed patch above ? I proposed it so as to get some feedback, but I would like it to be tested before deciding to merge it. Thanks!

Verified on git tag v2.6.0, environment variables on the parent process are passed to the external-check environment.

**Comment by C-Aniruddh** [comment]
@wtarreau @nyxi, you have an example of the `import-env` in use?

**Comment by nyxi** [comment]
@wtarreau anything else I can do to help with this?

**Comment by wtarreau** [comment]
Sorry, it seems I missed your response a few months ago :-(
I need to get back to this context now. I'm seeing you've tested and are satisfied with it, it's already a good step forward.

**Comment by wtarreau** [comment]
Hey @nyxi, @C-Aniruddh, @jvincze84, surprise! Now merged into 2.9. As mentioned in the commit message, the patch is simple enough that it could possibly be backported, depending on your use case, so please just let me know. We'll just wait a bit before backporting anything for now as we've had a huge 2.8 recently and want to wait a bit before knowing if another one is needed or not.

**Comment by wtarreau** [comment]
BTW, I forgot to say, as the commit message says, finally the keyword is "preserve-env" so that we can keep "import-env" free for a later usage if we want to selectively name certain variables for example.

**Comment by TheOptimisticFactory** [comment]
> Hey @nyxi, @C-Aniruddh, @jvincze84, surprise! Now merged into 2.9. As mentioned in the commit message, the patch is simple enough that it could possibly be backported, depending on your use case, so please just let me know. We'll just wait a bit before backporting anything for now as we've had a huge 2.8 recently and want to wait a bit before knowing if another one is needed or not.

This caused a regression: using `external-check` (without any argument) causes the following error to be thrown: 
```
[NOTICE]   (1) : haproxy version is 2.9.0-fddb8c1
[ALERT]    (1) : config : parsing [<config>/haproxy.cfg:17] : 'external-check' only supports 'preserve-env' as an argument, found ''.
[ALERT]    (1) : config : Error(s) found in configuration file : <config>/haproxy.cfg
[ALERT]    (1) : config : Fatal errors found in configuration.
```
- Related issue: https://github.com/haproxy/haproxy/issues/2380

**Comment by capflam** [comment]
Indeed, that was fixed. This will be shipped with the 2.9.1 (no ETA for now).
